### PR TITLE
Add knative backend (deploy and undeploy TODO)

### PR DESCRIPTION
**Pull request by boramalper**
Signed-off-by: Bora M. Alper <bora@boramalper.org>

### CODE REVIEW

File: driver.go [hunk 1]
```diff
@@ -191,6 +191,15 @@ func (d *Driver) run() {
 		lBackend.Deploy()
 	}
 
+	if runningInKnative() {
+		d.executor.(*knativeExecutor).Start()
+	}
+	if _, ok := d.executor.(*knativeExecutor); ok {
+		// TODO: not yet implemented
```
Comment (ustiugov) [inline driver.go:199]: ok to leave it as is

File: knative.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	taskResult := loadTaskResult(resultPayload)
+
+	atomic.AddInt64(&job.bytesRead, int64(taskResult.BytesRead))
+	atomic.AddInt64(&job.bytesWritten, int64(taskResult.BytesWritten))
+
+	return err
+}
+
+func (k *knativeExecutor) Deploy() {
+	panic("NOT YET IMPLEMENTED")
```
Comment (ustiugov) [inline knative.go]: use log.Fatal instead

File: knative.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func (k *knativeExecutor) Undeploy() {
+	panic("NOT YET IMPLEMENTED")
```
Comment (ustiugov) [inline knative.go]: use log.Fatal instead

File: knative.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	if err := s.ListenAndServe(); err != nil {
+		log.Fatal("http server failed: ", err)
+	}
+}
+
+func (k *knativeExecutor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		log.Error("failed to read http body: ", err)
```
Comment (ustiugov) [inline knative.go]: capitalize the first letter in a log message

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [x] import ""github.com/bcongdon/corral" instead of ease-lab, use `replace` in go.mod files to patch it up.

**Comment by boramalper** [comment]
@ustiugov Ready to merge (if we want to, or we can keep it as-is in the current branch as a fork). I have fixed some critical bugs and developed a CI test to make sure it does not go undetected again. I'll now focus on making it work on stock Knative.

File: examples/word_count/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FROM golang:1.16.6-buster AS builder
```
Comment (ustiugov) [inline examples/word_count/Dockerfile]: use the vhive golang image here

File: examples/word_count/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+WORKDIR /corral
+COPY . .
+WORKDIR /corral/examples/word_count
+RUN make
+
+FROM debian:buster
```
Comment (ustiugov) [inline examples/word_count/Dockerfile]:
```suggestion
FROM debian:buster-slim
```
smaller image

File: knative.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	var result taskResult
+	err := json.Unmarshal(payload, &result)
+	if err != nil {
+		log.Errorf("%s", err)
+	}
+	return result
+}
+
+func (k *knativeExecutor) invoke(payload []byte) (outputPayload []byte, err error) {
+	url := fmt.Sprintf("http://%s.default.127.0.0.1.nip.io:31080/", k.serviceName)
```
Comment (ustiugov) [inline knative.go:160]: should not it be *svc.local* instead of 127.0.0.1?

**Comment by ustiugov** [comment]
I think merging is fine but please fix few minor comments

File: knative.go (hunk 7 above)
Comment (boramalper) [reply knative.go:160]: Driver is a local program at the moment, and your suggestion makes sense only if the driver too is a K8s/Knative service. In that case, I won't be able to run the CI tests, so I was thinking of parameterising the URL in the future. Are you sure you want this change now?

Comment (ustiugov) [reply knative.go:160]: Up to you, later is ok

Comment (boramalper) [reply knative.go:160]: Let's do it later then

**Review by ustiugov — APPROVED** [review]

**Pull request merged**